### ISSUE TITLE
[binding request] APSystems Micro-Inverter EZ1-M

### ISSUE THREAD

**Issue by LukaNoah** [openhab/openhab-addons#16449]
APSystems has provided the EZ1 microinverters with a local API (description attached). With a binding, you would have the inverter as a thing and would not have to update the items with scripts. Unfortunately it doesn't work with the HTTP binding because json with multiple values are returned. There is already a HomeAssistant integration...
For beginners it would be perfect to have a binding.

Best regards
Heiko

[1701255814508-apsystems-ez1-local-api-user-manual.pdf](https://github.com/openhab/openhab-addons/files/14388466/1701255814508-apsystems-ez1-local-api-user-manual.pdf)

**Comment by deibich** [comment]
I'm also interested in the local API. Is anyone already developing a binding? If not, I'll start developing one in the next few days. The API description looks very straightforward, and it should be possible to get something up and running quickly.

**Comment by deibich** [comment]
Hey,
made a first version.

Source & Readme:  https://github.com/deibich/openhab-addons-apsystemsez1/tree/APsystemsEZ1/bundles/org.openhab.binding.apsystemsez1
Release: https://github.com/deibich/openhab-addons-apsystemsez1/releases

I'll create a topic in the forum and make the addon available in the marketplace over the weekend, if I find the time.

**Comment by 2relativ** [comment]
> Hey, made a first version.
> 
> Source & Readme: https://github.com/deibich/openhab-addons-apsystemsez1/tree/APsystemsEZ1/bundles/org.openhab.binding.apsystemsez1 Release: https://github.com/deibich/openhab-addons-apsystemsez1/releases
> 
> I'll create a topic in the forum and make the addon available in the marketplace over the weekend, if I find the time.

Thank you :) Just installed my new inverter and this binding. Will test it. So far, no issues. How can I support this to go from marketplace to an official binding?

**Comment by deibich** [comment]
> > > Hey, made a first version.
> > > Source & Readme: https://github.com/deibich/openhab-addons-apsystemsez1/tree/APsystemsEZ1/bundles/org.openhab.binding.apsystemsez1 Release: https://github.com/deibich/openhab-addons-apsystemsez1/releases
> > > I'll create a topic in the forum and make the addon available in the marketplace over the weekend, if I find the time.
> > 
> > 
> > Thank you :) Just installed my new inverter and this binding. Will test it. So far, no issues. How can I support this to go from marketplace to an official binding?
> 
> Did a short quick scan, allready looks solid. You might want to adjust the thing-type.xml with labels to 2-3 words < 23 chars and capitalized. Most important are lower-case-hypen naming convention for channel names as that would be breaking if you move it from marketplace to openhab itself. Good to see that is allready done!

Thank you for taking a look at the code. I will change the labels and submit a pull request soon.

**Comment by 2relativ** [comment]
![Screenshot_2024-05-26-08-01-16-88_fcd1c5899858d01174ac4962be56f19e.jpg](https://github.com/openhab/openhab-addons/assets/34840792/3eb3d528-76a5-4941-8292-f73e16db15da)

Moin, was mir noch so aufgefallen ist und ich nicht ganz nachvollziehen kann. Ab und an sind bei Energy Lifetime diese Einbrüche. Aber das ist so kurz, dass ich da nicht einmal drauf zeigen kann. Ich bin mir auch noch nicht sicher, woher das kommt. Ob vom Inverter, meinem Setup oder der API.

**Comment by lsiepel** [comment]
> was mir noch so aufgefallen ist und ich nicht ganz nachvollziehen kann. Ab und an sind bei Energy Lifetime diese Einbrüche. Aber das ist so kurz, dass ich da nicht einmal drauf zeigen kann. Ich bin mir auch noch nicht sicher, woher das kommt. Ob vom Inverter, mei

English please.

**Comment by 2relativ** [comment]
![Screenshot_2024-05-26-08-01-16-88_fcd1c5899858d01174ac4962be56f19e.jpg](https://github.com/openhab/openhab-addons/assets/34840792/3eb3d528-76a5-4941-8292-f73e16db15da)

Ah, shoot. Sorry. 😅 Too early in the morning. What I was saying is that I noticed some drops in the total energy channel. I see the same drops on each panel. Does anyone else notice this? Or is my setup faulty?

**Comment by 2relativ** [comment]
Another little feedback: At least while on mobile it is irritating that channels are labeled the same by default. Example "Energy lifetime". You can link both panels and the device itself and you can't really tell on mobile. Maybe change the labeling a little bit?

**Comment by 2relativ** [comment]
> ![Screenshot_2024-05-26-08-01-16-88_fcd1c5899858d01174ac4962be56f19e.jpg](https://github.com/openhab/openhab-addons/assets/34840792/3eb3d528-76a5-4941-8292-f73e16db15da)
> 
> Ah, shoot. Sorry. 😅 Too early in the morning. What I was saying is that I noticed some drops in the total energy channel. I see the same drops on each panel. Does anyone else notice this? Or is my setup faulty?

Regarding this with the drop ins... I set the pull intervall to 20 seconds. I oriented myself by the shelly binding which also has 20 seconds as default. Now the graph looks nicer 😁👍🏻 maybe set the default to 20 seconds?

**Comment by deibich** [comment]
> Ah, shoot. Sorry. 😅 Too early in the morning. What I was saying is that I noticed some drops in the total energy channel. I see the same drops on each panel. Does anyone else notice this? Or is my setup faulty?

> Regarding this with the drop ins... I set the pull intervall to 20 seconds. I oriented myself by the shelly binding which also has 20 seconds as default. Now the graph looks nicer 😁👍🏻 maybe set the default to 20 seconds?

I see those drops too. But don't know why. The binding should just forwards the response from the inverter to the channels. I expect the inverter to response with invalid data.

Each response from the inverter comes with a `message` which can be `SUCCESS`or `FAILED`. The API does not describe the key in detail. It might indicate whether the request was successful or not. At the moment The binding does not check the value of this key. 

You can set the log of `org.openhab.binding.apsystemsez1` to `DEBUG` to analyze the response from the inverter if you like. The log contains the key value pair `message`. 

I'll review my log in the coming days.

**Comment by deibich** [comment]
> Another little feedback: At least while on mobile it is irritating that channels are labeled the same by default. Example "Energy lifetime". You can link both panels and the device itself and you can't really tell on mobile. Maybe change the labeling a little bit?

Yeah this comes from the reuse of the same channel-types. I'm not sure if same channel-type can get different descriptions in different channel-groups. I'll check it out. Otherwise different channel-types must be defined.

**Comment by deibich** [comment]
I have analyzed my logs, and the drops in the graph seem to be related to unsuccessful responses from the inverter.
For example:
```LOG
[DEBUG] [temsez1.internal.APsystemsEZ1Handler] - Response from endpoint /getOutputData: EZ1OutputDataResponseData [message=FAILED, deviceId=XXXXXX, data=Data [p1=0.0, e1=0.0, te1=0.0, p2=0.0, e2=0.0, te2=0.0]]
```
The binding now drops responses that are not successful (message is not `SUCCESS`). 

Occasionally, the inverter responds with a SUCCESS but leaves the data object empty. This happens to me exclusively with the `/getOnOff` endpoint.
```LOG
[TRACE] [temsez1.internal.APsystemsEZ1Handler] - Raw Response: {"data":{"status":""},"message":"SUCCESS","deviceId":"XXXXXX"}
[DEBUG] [temsez1.internal.APsystemsEZ1Handler] - Response from endpoint /getOnOff: EZ1OnOffResponseData [message=SUCCESS, deviceId=XXXXXX, data=Data [status=]]
```
`status` should be `1`or `0`

**Comment by Gehetzter** [comment]
First of all I really like to thank you for creating this binding!

I've got a request for improvement:
Could you please send the requests to the device only if the corresponding channels are linked?
I only use the power and energy values but see all 4 api requests in the log.

**Comment by deibich** [comment]
Thank you for your input.

> First of all I really like to thank @deibich for creating this binding!
> 
> I've got two requests for improvement:
> 
>     1. Could you please send the requests to the device only if the corresponding channels are linked?
>        I only use the power and energy values but see all 4 api requests in the log.
>        Sometimes the getOnOff request fails and I get a warning in my log then but I don't even use it.
>        Maybe there are people running this thing over a mobile network with limited data volume. Could benefit too.

Currently the binding just polls all five endpoints. This was easier to start with in the first place. Since there is not much data send and I use the binding in my home WIFI this was not considered by me.

A simple check if the corresponding Channel is linked should be easy doable.

>     2. It would be great to retry a failed request. Maybe for a configurable amount in advanced settings?

Do you talk about a failed request in terms of "network failure"? Currently the Thing goes offline if a request fails. But the binding sends a request every "refreshIntervalSeconds" and comes back online if the inverter is available then.

**Comment by Gehetzter** [comment]
> Do you talk about a failed request in terms of "network failure"? Currently the Thing goes offline if a request fails. But the binding sends a request every "refreshIntervalSeconds" and comes back online if the inverter is available then.

No, I don't mean a network failure, even though I don't know what's causing this in my case. In my setup sometimes a response from the inverter doesn't contain the SUCCESS message, but I'm sure the inverter sends the data like this. Data is then dropped by the binding, which is great, but an instant retry would be great, to not miss a point in time, e.g. when I pull every minute and one response is dropped I'm missing the data of this specific minute. Wouldn't care if the data comes in correct a bit off in time because of the retry.

**Comment by Gehetzter** [comment]
> So you experience the same as described in a previous comment: [#16449 (comment)](https://github.com/openhab/openhab-addons/issues/16449#issuecomment-2156739376) , too?
> 
Yes, you're right. I haven't noticed that it's exclusively the /getOnOff, but you could be right.
> Should be possible to implement a simple retry. Also a second parameter with a configurable timeout for a failed request as a advanced setting, like you already described, would be nice.
> 
> I was a quiet busy in the past but think I could implement the basic functionality in the near future.
I would be very pleased if you could include it to your roadmap, no need to rush, your private life comes first. As a workaround I've just changed the log level of the binding to get rid of the entries in frontail...

**Comment by ned14** [comment]
Apologies for resurrecting this issue, however the EZ1-M bindings repo doesn't have an enabled issue tracker.

Could support for the new REST endpoints in recent firmwares be added? Here they are:

curl -s http://192.168.xxx.xxx:8050/getOutputDataDetail
{"data":{"c1":2.15280,"v1":50.82336,"p1":101,"e1":0.13143,"te1":0.85877,"c2":2.14110,"v2":50.66208,"p2":100,"e2":0.13945,"te2":0.58355,"gf":49.91000,"gv":242,"t":29},"message":"SUCCESS","deviceId":"E0701xxxxxxx"}%   
MPPT 1:
c1 = amperes
v1 = Volt
p1 = Watt
e1 = kWh (today)
te1 = kWh (total since the last reset)
MPPT 2:
c2 = amperes
v2 = Volt
p2 = Watt
e2 = kWh (today)
te2 = kWh (total since the last reset)
Mains grid:
gf = Hertz
gv = volt
t = inverter temperature

The big gain here is detailed insights into MPTT voltage and amps, plus grid voltage and the inverter internal temperature.

curl -s http://192.168.xxx.xxx:8050/getDefaultMaxPower        
{"data":{"power":"200"},"message":"SUCCESS","deviceId":"E0701xxxxxxx"}

The DefaultMaxPower is what the inverter sets MaxPower to on boot, and it is stored in flash. setDefaultMaxPower lets you write that flash value, and it appears to refuse further writes for several minutes.

**Comment by lsiepel** [comment]
Maybe @deibich can consider to release it in the marketplace ? Low entry way of sharing the binding and getting feedback.

**Comment by deibich** [comment]
Thank you for pointing out the update. Is there any more information available about it? 
On https://emea.apsystems.com/document-library/ I can only find the documentation for v1.1, which does not yet include the changes. Do you know from which firmware version the new endpoints will be available?

**Comment by ned14** [comment]
> Thank you for pointing out the update. Is there any more information available about it? On https://emea.apsystems.com/document-library/ I can only find the documentation for v1.1, which does not yet include the changes. Do you know from which firmware version the new endpoints will be available?

These new endpoints appear to have been shipping in firmwares since last year.

I would suggest try probing for getOutputDataDetail, if it doesn't exist, fallback to the less detailed API.

The DefaultMaxPower endpoint is probably best left out for now, as its value added is low whereas for getOutputDataDetail there are several values very useful to record over time (esp MPTT voltage).
